go-kit/metrics/provider/otel/provider.go

```diff
@@ -3,371 +3,164 @@ package otel
-	"go.opentelemetry.io/otel/sdk/export/metric/aggregation"
-	"go.opentelemetry.io/otel/sdk/metric/aggregator/histogram"
-	metriccontroller "go.opentelemetry.io/otel/sdk/metric/controller/basic"
-	processor "go.opentelemetry.io/otel/sdk/metric/processor/basic"
+	"go.opentelemetry.io/otel/sdk/metric"
 	"go.opentelemetry.io/otel/sdk/resource"
 
 	xmetrics "github.com/heroku/x/go-kit/metrics"
-	"github.com/heroku/x/go-kit/metrics/provider/otel/selector/explicit"
 )
 
-var _ metrics.Counter = (*Counter)(nil)
-var _ metrics.Gauge = (*Gauge)(nil)
-var _ metrics.Histogram = (*Histogram)(nil)
-
 const (
-	// The values of these attributes should be the service name.
-	serviceKey     = "_service"
-	serviceNameKey = "service.name"
-	componentKey   = "component"
-
-	// The "service.namespace" attribute should be the cloud/deploy (e.g. "va01").
-	serviceNamespaceKey = "service.namespace"
-
-	// The "service.instance.id" attribute will be an identifier for this specific instance of the service (e.g. "web.1").
-	serviceInstanceIDKey = "service.instance.id"
+	defaultExponentialHistogramMaxSize  = 160
+	defaultExponentialHistogramMaxScale = 20
+)
 
-	// The values of these attributes should be the stage (e.g. "production").
-	stageKey      = "stage"
-	subserviceKey = "_subservice"
+var DefaultReaderInterval = time.Minute
 
-	// The value of the "cloud" attribute should be the cloud (e.g. "heroku.com")
-	cloudKey = "cloud"
-)
+type config struct {
+	ctx                 context.Context // used for init and shutdown of the otlp exporter and other bits of this Provider
+	serviceResource     *resource.Resource
+	prefix              string
+	collectPeriod       time.Duration
+	temporalitySelector metric.TemporalitySelector
+	aggregationSelector metric.AggregationSelector
+	exporterFactory     exporterFactory
+}
 
 // Provider initializes a global otlp meter provider that can collect metrics and
 // use a collector to push those metrics to various backends (e.g. Argus, Honeycomb).
 // Initialize with New(...). An initialized Provider must be started before
 // it can be used to provide a Meter (i.e. p.Start()).
 type Provider struct {
-	ctx                 context.Context // used for init and shutdown of the otlp exporter and other bits of this Provider
-	serviceNameResource *resource.Resource
-	optionCache         explicit.OptionCache
-	selector            metricexport.AggregatorSelector
-	exporter            exporter
-	controller          controller
-
-	defaultTags   []string
-	prefix        string
-	collectPeriod time.Duration
+	cfg           *config
+	meterProvider *metric.MeterProvider
+	viewCache     *viewCache
 
 	mu         sync.Mutex
 	counters   map[string]*Counter
 	gauges     map[string]*Gauge
 	histograms map[string]*Histogram
 }
 
+type viewCache struct {
+	lock    sync.RWMutex
+	streams map[string]metric.Stream
+}
+
+func (v *viewCache) Store(stream metric.Stream) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	v.streams[stream.Name] = stream
```

Comment:
Could there be some concurrency issue here because we're not doing a read-before-write?

Reply:
I don't think so. There is a higher level lock that happens at the `newHistogram` that would prevent concurrent metric creation calls. There is also a caching layer there that will return an existing histogram if one of the same name has already been created.